### ISSUE TITLE
Program crashes with constant tensors

### ISSUE THREAD

**Issue by gbenfis** [Dobiasd/frugally-deep#442]
My program crashes when loading the model from the json file. Upon debugging, it crashes when looking for a keras_history in a node, and not finding any attribute with that name:

    Assertion failed: it != m_data.m_value.object->end(), file C:\Users\e1081782\dependency\json\single_include\nlohmann\json.hpp, line 21449

This is most probably linked to the fact that my model normalizes the input tensor by subtracting a constant tensor (which has no history). Is there any workaround to make it execute?

        {
          "module": "keras.src.ops.numpy",
          "class_name": "Subtract",
          "config": { "name": "subtract_2" },
          "registered_name": "Subtract",
          "name": "subtract_2",
          "inbound_nodes": [
            {
              "args": [
                {
                  "class_name": "__keras_tensor__",
                  "config": {
                    "shape": [ null, 11 ],
                    "dtype": "float32",
                    "keras_history": [ "input_1", 0, 0 ]
                  }
                },
                {
                  "class_name": "__tensor__",
                  "config": {
                    "value": [ [ 0.014991031028330326, -5.773593159119628e-08, 0.3049706220626831, 0.9997414946556091, 0.04997607693076134, 1.0001437664031982, 0.9999415874481201, 1.0001988410949707, 1.0000635385513306, 2.4990406036376953, 0.4999502897262573 ] ],
                    "dtype": "float32"
                  }
                }
              ],
              "kwargs": {}
            }
          ]
        },

**Comment by Dobiasd** [comment]
Thanks for reporting!

Could you upload your model, so I can reproduce the error locally and debug it?

**Comment by gbenfis** [comment]
[fdeep_model.json](https://github.com/user-attachments/files/19896036/fdeep_model.json)

It doesn't let me upload .keras files here. I've uploaded the json file in case it can help!

**Comment by gbenfis** [comment]
Sure, here it is: https://drive.google.com/file/d/1bZPpGPowOxmEW1qH16-7hV5Xgkh8ty0L/view?usp=drivesdk

Thanks!

**Comment by Dobiasd** [comment]
Thanks. With this, I can [reproduce](https://gist.github.com/Dobiasd/0d8a0b66a4c4d1ba364410b5674882b3) the problem locally too. 👍

Can you create a minimal Python source code that builds a model running into that problem? (no history, likely from the constant tensor)

**Comment by gbenfis** [comment]
Sure, here it is. I think the problem boils down to the normalization layer.

        import tensorflow as tf
        import numpy as np
        
        x = np.array([[0.1, 0.2]])
        inputs = tf.keras.Input(shape=(2,), name="input_1")
        input_norm_layer = tf.keras.layers.Normalization()
        input_norm_layer.adapt(x)
        outputs = (inputs - input_norm_layer.mean) / tf.sqrt(input_norm_layer.variance)
        model = tf.keras.Model(inputs=inputs, outputs=outputs)
        model.save('test.keras')

**Comment by Dobiasd** [comment]
Thanks a lot. This works nicely to reproduce the problem. I'll see if I can fix it and get back to you here. 🧑‍🔧

**Comment by gbenfis** [comment]
Hi,

So I found a way around this problem by avoiding doing the normalization manually as I did above, but letting the normalization layer handle it:

        import tensorflow as tf
        import numpy as np
        
        x = np.array([[0.1, 0.2], [0.5, 0.7]])
        y = np.array([[0.1, 0.15], [0.2, 0.35]])
        inputs = tf.keras.Input(shape=(2,), name="input_1")
        input_norm_layer = tf.keras.layers.Normalization()
        input_norm_layer.adapt(x)
        outputs = input_norm_layer(inputs)
        output_norm_layer = tf.keras.layers.Normalization(invert=True)
        output_norm_layer.adapt(y)
        outputs = output_norm_layer(outputs)
        model = tf.keras.Model(inputs=inputs, outputs=outputs)
        model.save('test.keras')

Now the previous error disappears but I get this new one (which seems linked to the inverse normalization layer):

    FAILURE test failed: output=0 pos=0,0,0 value=-45.8373 target=0.0429067 

Any idea why?

**Comment by Dobiasd** [comment]
Awesome that you found a workaround! Implementing support for constant-value layers in frugally-deep is non-trivial. Having you unblocked by this solution takes away the pressure. 😌

Your current problem is way simpler. Frugally-deep does not yet support `invert=True` for `Normalization` layers. I'll add this support and get back to you.

**Comment by gbenfis** [comment]
Sounds great, thank you!

**Comment by Dobiasd** [comment]
Ok, support for `Normalization` layers with `invert=True` is implemented now and available in the [new version](https://github.com/Dobiasd/frugally-deep/releases/tag/v0.18.1) (`0.18.1`). ✅
Please let me know if it works for your particular problem too.

And thanks for these great reports, helping to make frugally-deep a bit better. ❤

I'll keep this issue open nonetheless, to have a public todo item for implementing support for constant-value layers. 🧑‍💼

**Comment by gbenfis** [comment]
Hi, I have now tested it and it works perfectly! 
Thanks a lot for the super quick support!
